File: valid_anagram/valid_anagram.ts
// My Solution
const isAnagram = (s: string, t: string) => {
  let tracker = {};
  let passing = true;
  for (let char of s) {
    if (char in tracker) {
      //if(Object.keys(tracker).includes(char)) is obsolete
      tracker[char] += 1;
    } else {
      tracker[char] = 1;
    }
  }
  for (let char of t) {
    if (char in tracker) {
      tracker[char] -= 1;
    } else {
      passing = false;
    }
  }
  if (!Object.values(tracker).every((value) => value === 0)) {
    passing = false;
  }
  return passing;
};

//Optimal Solution
var isAnagramOptimal = function (s, t) {
  if (s.length != t.length) return false;

  const charMap = Array(26).fill(0);

  for (let i = 0; i < s.length; i++) {
    charMap[s.charCodeAt(i) - "a".charCodeAt(0)]++;
    charMap[t.charCodeAt(i) - "a".charCodeAt(0)]--;
  }

  return charMap.every((count) => count === 0);
};

//Using Maps
var isAnagramUsingMap = function (s, t) {
  if (s.length != t.length) return false;

  const charMap = Array(26).fill(0);

  for (let i = 0; i < s.length; i++) {
    charMap[s.charCodeAt(i) - "a".charCodeAt(0)]++;
    charMap[t.charCodeAt(i) - "a".charCodeAt(0)]--;
  }

  return charMap.every((count) => count === 0);
};

File: valid_anagram/valid_anagram.py
# My Solution
def isAnagram(self, s: str, t: str) -> bool:
    tracker = {}
    for letter in s:
        if letter in tracker:
            tracker[letter] += 1
        else:
            tracker[letter] = 1
    for letter in t:
        if letter in tracker:
            tracker[letter] -= 1
        else:
            return False
    if all(value == 0 for value in tracker.values()):
        return True
    else:   
        return False


# Optimal Solution
def isAnagram(self, s: str, t: str) -> bool:
    return Counter(s) == Counter(t)
